### ISSUE TITLE
[Feature] iPlayer WWW player

### ISSUE THREAD

**Issue by t0mtee** [jurialmunkey/plugin.video.themoviedb.helper#839]
### Describe the feature that you'd like

A player for the iPlayer WWW addon would be great

### Why is this feature needed? What problem does it solve?

iPlayer WWW has no feature to export items to your library, and this would make it possible to play it directly from there. I understand that the author of this addon has no obligation to fill in for the gaps of another, however I have been directed here from there. I've tried to make my own player, however I'm not knowledgable enough in the way Kodi addons work and all attempts haven't worked.

### Checklist

- [X] I have searched the issues and this feature has not been requested.
- [X] I have updated TMDbHelper to the most recent version.

**Comment by jurialmunkey** [comment]
I dont think this is possible for this plugin.

The search is very slow, which means it will take forever to play something while you wait for the search. Plus the metadata for the items is very limited so there aren't many options to match on in terms of season or episode numbers. 

That means you have to rely on numbers in the labels which dont really follow any consistent pattern Some shows have numbers for episodes, others just titles, some have all episodes in one big folder and others have them split into series.

Without any reliable metadata to match on it you will get a lot of incorrect matches.

**Comment by jurialmunkey** [comment]
Actually might have got something working. It is difficult for me to test though as I don't live in the UK.

@OsoaGH - please try v4.11.7 which will give you a "BBC iplayer www" player for episodes only (not sure if this plugin even does movies but that's probably going to be too difficult).

It is very rough so I expect it wont be able to match some things. I tested with the show Snowfall and was at least able to get matches with that fairly well.

Basically what the player does is 

1. Goes into "Programme List A-Z"
2. Selects the letter the show starts with
3. Looks for a label that contains the show name AND has the text "available" at the end
4. Looks for a label that starting with the format "X. Title" (where X is episode number) and plays if found.
5. If there was no match, it assumes that we're in the wrong series number, so it looks for a folder named "Series X" (where X is the season number).
6. Repeats step 4 in that Series folder and plays if found (else player fails).

As I said, very rough, but there really isn't much to work with in this plugin. Need to go via the A-Z list as it is much faster than searching (the search speed is abysmal).

@berkhornet - This may interest you, as I know you use iplayerwww plugin too.

**Comment by t0mtee** [comment]
Ok nice, this seems to work well. I've tested with three shows: Doctor Who, Danger Mouse and Peaky Blinders. It worked well with Doctor Who and Danger Mouse, all episodes matched, however on Peaky Blinders, it failed. I think this because the first two have specific episode names while Peaky Blinders is just named "Episode 1", "Episode 2", and thus doesn't follow the "X. Title" format. Example:
Doctor Who on the addon
![image](https://user-images.githubusercontent.com/33041004/188266229-e79214d2-16c4-4d8e-93fd-12b8125490a2.png)
Peaky Blinders on the addon
![image](https://user-images.githubusercontent.com/33041004/188266242-c27371c8-c76e-4126-9442-d81026afd556.png)

Maybe for step 4, test for both formats? Thanks.

**Comment by t0mtee** [comment]
Also for films, I think it may be possible. There is a menu that is very similar to the A-Z Programmes one for Films in Categories > Films.

**Comment by jurialmunkey** [comment]
> Maybe for step 4, test for both formats? Thanks.

Won't work if the title is just "Episode X" because there's no way to determine what season it is. You could be playing Episode 1 of any of Seasons 1-6. The iplayer plugin opens in a different season depending on whichever was last opened, so we can't even assume that first folder is always Season 1. It *might* be Season 1 but it could also be any of Season 2-6.

That's why the matching is particularly difficult when the end plugin doesn't include any season or episode metadata. Trying to go entirely off the label when we don't even know what season we're in is practically impossible.

**Comment by t0mtee** [comment]
I can try and reach out to the addon dev to see if they can add a setting to display the season next to the episode title if that would help

**Comment by jurialmunkey** [comment]
> I can try and reach out to the addon dev to see if they can add a setting to display the season next to the episode title if that would help

If they've got the actual season and episode numbers then it would be better to have it added correctly to ListItem.Season and ListItem.Episode rather than just the label.

**Comment by t0mtee** [comment]
> > I can try and reach out to the addon dev to see if they can add a setting to display the season next to the episode title if that would help
> 
> If they've got the actual season and episode numbers then it would be better to have it added correctly to ListItem.Season and ListItem.Episode rather than just the label.

Looking at the commits of the addon, it looks like they may be adding ListItem.Episode (I may be mistaken however) for the next update. Take a look yourself: [commits](https://github.com/vonH/plugin.video.iplayerwww/commits/matrix). I'll direct the dev to this issue, hopefully they will take a look and make the changes it needs.

**Comment by CaptainTK** [comment]
Unfortunately, all that plugin.video.iplayerwww does is parse the BBC website. The website does not provide structured data on series and episode numbers. They are simply contained in the title or subtitle. Strangely enough, for Peaky Blinders I can either find pages that show "Series X - Y. Title" or "Episode Y", but not "Series X - Y. Episode".

In any case, since there is no logic to the titles on the BBC side, it would be equally hard for the plugin.video.iplayerwww to extract season and episode information correctly from the metadata on the website as it would be for plugin.themoviedb.helper.

In short: Neither should fix this, the BBC should. But they most likely have no interest in doing so.

**Comment by jurialmunkey** [comment]
@CaptainTK thanks for this info. I figured as much really. We have a similar situation in Australia for our national broadcasters catchup service ABC iView. It has very slightly better metadata to go off from the website but generally speaking has pretty much the same issue where there isn't a clear structure to season folders and episode naming/numbering varies depending on show.

**Comment by jurialmunkey** [comment]
> 
> Looking at the commits of the addon, it looks like they may be adding ListItem.Episode (I may be mistaken however) for the next update. Take a look yourself: [commits](https://github.com/vonH/plugin.video.iplayerwww/commits/matrix).

@OsoaGH that's just setting the mediatype for each listitem to episode. Which is excellent news for skins as it helps skins determine more accurately what type of content each item is, but doesn't really help us here in terms of the inconsistent numbering the website itself provides.

**Comment by t0mtee** [comment]
ah, ok. honestly, i'm happy with the player right now as it does everything i need it to do (don't have any shows that follow the peaky blinders format right now) so feel free to close it if you like. if there's any later developments that make the peaky blinders format possible then we could reopen

**Comment by jurialmunkey** [comment]
I'll have a look at the movies section in a bit and see if it is feasible and can cobble together something for that before closing this issue. 

I don't think there's a solution for the peaky blinders numbering though (other than the website itself changing). Might be able to go to a select dialog in that instance but possibly might not work due to how the series folders are nested.

**Comment by jurialmunkey** [comment]
Okay when I make the next release the above commit will be included to add support for movies. It will also give an additional "Search BBC iPlayer" function for episodes, which will give you a dialog pop-up to select with for those cases where the main "Play with" function doesn't work.

Also since movies are matching on title only, it isn't possible to differentiate movies with the same title. e.g. The 2018 "A Star is Born" is available on iplayer but this will also match if you try to play any of the five different film adaptations of this screenplay. Same goes for remakes like Total Recall or Robocop. 

You'll also get false positives where the title makes up part of the name of another film e.g. playing "Terminator" when only "Terminator Genisys" is available will play Terminator Genisys. I can't do stricter matches due to the way the website might add in extra information on the end of the title or how some titles add in text formatting tags. Generally it should be okay but there are definitely going to be false matches.

**Comment by t0mtee** [comment]
This is great, thanks so much.

**Comment by berkhornet** [comment]
> Actually might have got something working. It is difficult for me to test though as I don't live in the UK.
> 
> @OsoaGH - please try v4.11.7 which will give you a "BBC iplayer www" player for episodes only (not sure if this plugin even does movies but that's probably going to be too difficult).
> 
> It is very rough so I expect it wont be able to match some things. I tested with the show Snowfall and was at least able to get matches with that fairly well.
> 
> Basically what the player does is
> 
> 1. Goes into "Programme List A-Z"
> 2. Selects the letter the show starts with
> 3. Looks for a label that contains the show name AND has the text "available" at the end
> 4. Looks for a label that starting with the format "X. Title" (where X is episode number) and plays if found.
> 5. If there was no match, it assumes that we're in the wrong series number, so it looks for a folder named "Series X" (where X is the season number).
> 6. Repeats step 4 in that Series folder and plays if found (else player fails).
> 
> As I said, very rough, but there really isn't much to work with in this plugin. Need to go via the A-Z list as it is much faster than searching (the search speed is abysmal).
> 
> @berkhornet - This may interest you, as I know you use iplayerwww plugin too.

@jurialmunkey thanks - just picked this up. I've not used the player function previously but will take it for a test drive!
